docs/design-proposals/2023-11 TraceQL Metrics.md

```diff
@@ -0,0 +1,112 @@
+---
+Authors: Martin Disibio (@mdisibio)
+Created: 2023 November
+Last updated: 2023-11-13
+---
+
+# TraceQL Metrics
+
+## Summary
+
+This design document describes TraceQL language additions for extracting time-series metrics like error rates and latencies from traces.   
+
+This document is **not** meant to be a complete language specification. Rather it is meant to invite the community to comment on the direction and capabilities.
+
+### Table of Contents
+
+- [Intro](#intro)
+- [First Stage](#first-stage)
+  - [Interval](#interval)
+  - [Grouping](#grouping)
+  - [Filtering](#filtering)
+- [Additional Stages](#additional-stages)
+- [Arithmetic](#arithmetic)
+- [Notes](#notes)
+  - [Step Interval](#step-interval)
+
+## Intro
+
+Let's look at an example straight away.  This query plots the error rate by service over time:  
+`{status = error } | rate() by (resource.service.name)`
+
+A normal TraceQL query finds matching spans and returns data points like time, duration, service name, status code, etc.  A metrics query at the core does the same thing: _find matching spans_, but instead of seeing individual results, they are aggregated over time to create time-series.  A metrics query is built by extending a regular query with metrics pipeline stages.  The structure looks like this:
+
+`<span query...> | <metrics query...>`
+
+An important understanding is that _any_ valid TraceQL query can be turned into a metrics query. This finds the call rates to service B but only when initiated upstream by service A:  
+`{resource.service.name="A"} >> {resouce.service.name="B"} | rate()`
+
+
+## First Stage
+
+The first stage in a metrics query is responsible for turning spans into time series. This is a unique intersection between the two data types and has specialized aggregation functions.  We use the `..._over_time()` suffix to distinguish them from their counterparts.
+
+| function                                   | description                |
+---------------------------------------------| ---------------------------- 
+|`rate()`                                    | The rate of spans per second
+|`count_over_time()`                         | The count of spans
+|`avg_over_time(<field>)`                    | The average of a numeric value like `duration` or [http.request.body.size](https://opentelemetry.io/docs/specs/semconv/attributes-registry/http/)
+|`max_over_time(<field>)`                    | The maximum of a numeric value.
+|`min_over_time(<field>)`                    | The minimum of a numeric value.
+|`quantile_over_time(<field>, q1, q2, ...)`  | The quantile of a numeric value (e.g. p95). Multiple quantiles can be requested and a time-series is created for each. |
+
+### Interval
+All aggregations take an optional interval like `30s`, `5m`, `1h`. If not specified then the interval will automatically match the step interval of the query. The interval is the last parameter of each function.
+
+`rate(5m)`  
```

Comment:
Wouldn't it be more prometheus-like to use `[]` to indicate interval? E.g. `rate[5m]()` or `quantile_over_time[5m](duration, 0.95)`. Then the arguments of the function wouldn't be mixed with interval

Reply:
Hi indeed we debated that internally too.  Our rationale is:  in prometheus `[5m]` is meaningful on its own without a specific function.  Both `metric{}` and `metric{}[5m]` are valid constructs (with different types).  That doesn't apply to Tempo because there are no underlying `metric` time-series.  They must be created by these aggregation functions and in that case there is no need to make the interval special, it could be simplified to an argument.  To clarify even further:  `{.http.status_code>=500}` is valid and finds spans. But `{.http.status_code>=500}[5m]` doesn't mean anything without the aggregation.

To tweak your example it would probably look like this:  `rate([5m])` and `quantile_over_time(duration, 0.95 [5m])`.